### PR TITLE
use CountedMap in pallet-bags-list

### PR DESCRIPTION
**Pull request by kianenigma**

### CODE REVIEW

File: frame/bags-list/src/list/mod.rs [hunk 1]
```diff
@@ -77,17 +77,18 @@ pub struct List<T: Config>(PhantomData<T>);
...
 impl<T: Config> List<T> {
 	/// Remove all data associated with the list from storage. Parameter `items` is the number of
-	/// items to clear from the list. WARNING: `None` will clear all items and should generally not
-	/// be used in production as it could lead to an infinite number of storage accesses.
+	/// items to clear from the list.
+	///
+	/// ## WARNING
+	///
+	/// `None` will clear all items and should generally not be used in production as it could lead
+	/// to a vary large number of storage accesses.
```
Comment (KiChjang) [inline frame/bags-list/src/list/mod.rs:85]:
```suggestion
	/// to a very large number of storage accesses.
```

**Review by KiChjang — APPROVED** [review]

File: frame/support/src/storage/types/counted_map.rs [hunk 2]
```diff
@@ -262,9 +263,10 @@ where
 	}
 
 	/// Remove all value of the storage.
-	pub fn remove_all() {
-		CounterFor::<Prefix>::set(0u32);
-		<Self as MapWrapper>::Map::remove_all(None);
+	pub fn remove_all(maybe_limit: Option<u32>) {
+		let leftover = Self::count().saturating_sub(maybe_limit.unwrap_or_else(Bounded::max_value));
```
Comment (kianenigma) [inline frame/support/src/storage/types/counted_map.rs:267]: this is the only logical change here.

**Review by shawntabrizi — APPROVED** [review]
we need to double check the storage prefix is the same?

**Comment by kianenigma** [comment]
> we need to double check the storage prefix is the same?

Code points to the prefix being `CouterFor`, added a test for it as well.

**Comment by kianenigma** [comment]
Let's see if this bad boy is still alive:

**Comment by kianenigma** [comment]
/try-runtime queue RUST_LOG=runtime=debug --execution Native --no-spec-name-check on-runtime-upgrade live --uri ws://kusama

**Comment by kianenigma** [comment]
anyways, locall I get: 

```
2021-11-10 09:46:07.016  INFO main runtime::bags_list: [10031084] 👜checked bags-list prefix to be correct and have 6753 nodes
```

**Comment by kianenigma** [comment]
bot merge

File: frame/support/src/storage/types/counted_map.rs (hunk 2 above)
Comment (thiolliere) [reply frame/support/src/storage/types/counted_map.rs:267]: this is not correct, the documentation of remove_all for the storage map is indeed missing, but underneath it calls [sp-io::storage::clear_prefix](https://paritytech.github.io/substrate/master/sp_io/storage/fn.clear_prefix.html)
Which deletes all keys in the overlay and up to limit in the backend.

Thus we cannot keep an accurate leftover here, we need to change sp-io::storage::clear_prefix to return here the number of key deleted in the overlay.

File: frame/bags-list/src/list/mod.rs [hunk 3]
```diff
@@ -77,17 +77,18 @@ pub struct List<T: Config>(PhantomData<T>);
...
+	/// ## WARNING
+	///
+	/// `None` will clear all items and should generally not be used in production as it could lead
+	/// to a very large number of storage accesses.
 	pub(crate) fn clear(maybe_count: Option<u32>) -> u32 {
 		crate::ListBags::<T>::remove_all(maybe_count);
+		let pre = crate::ListNodes::<T>::count();
 		crate::ListNodes::<T>::remove_all(maybe_count);
-		if let Some(count) = maybe_count {
-			crate::CounterForListNodes::<T>::mutate(|items| *items - count);
```
Comment (thiolliere) [inline frame/bags-list/src/list/mod.rs:86]: this previous code also doesn't seems correct, it is not ensure that `count` items are removed.

File: frame/support/src/storage/types/counted_map.rs (hunk 2 above)
Comment (KiChjang) [reply frame/support/src/storage/types/counted_map.rs:267]: @kianenigma wat happens nao?

Comment (thiolliere) [reply frame/support/src/storage/types/counted_map.rs:267]: follow up is https://github.com/paritytech/substrate/pull/10231

**Pull request merged**